index.js

```diff
@@ -1,205 +1,40 @@
-    }
+      domain = domain.toString()
 
-    e.statusCode = e.statusCode || 500
-    throw e
-  }
-}
-
-function getSecret(request, reply, cb) {
-  request
-    .jwtDecode({ decode: { complete: true } })
-    .then(decoded => {
-      const { header } = decoded
-
-      // If the algorithm is not using RS256, the encryption key is Auth0 client secret
-      if (header.alg.startsWith('HS')) {
-        return cb(null, request.auth0Verify.secret)
+      if (!domain.match(/^http(?:s?)/)) {
+        domain = new URL(`https://${domain}`).toString()
+      } else {
+        // Add missing trailing slash if it hasn't been provided in the config
+        domain = new URL(domain).toString()
       }
 
-      // If the algorithm is RS256, get the key remotely using a well-known URL containing a JWK set
-      getRemoteSecret(request.auth0Verify.domain, header.alg, header.kid, request.auth0VerifySecretsCache)
-        .then(key => cb(null, key))
-        .catch(cb)
-    })
-    .catch(cb)
-}
-
-async function authenticate(request, reply) {
-  try {
-    await request.jwtVerify()
-  } catch (e) {
-    for (const [jwtMessage, errorMessage] of fastifyJwtErrors) {
-      if (e.message.match(jwtMessage)) {
-        throw new Unauthorized(errorMessage, { a: 1 })
-      }
-    }
-
-    if (e.statusCode) {
-      throw e
+      options.jwksUrl = `${domain}.well-known/jwks.json`
+    } else if (!secret) {
+      // Throw missing options error here to prevent confusing error
+      // message from fastify-jwt-jwks being thrown to user
+      throw new Error(errorMessages.missingOptions)
     }
 
-    throw new Unauthorized(e.message)
-  }
-}
-
-function fastifyAuth0Verify(instance, options, done) {
-  try {
-    // Check if secrets cache is wanted - Convert milliseconds to seconds and cache for a week by default
-    const ttl = parseFloat('secretsTtl' in options ? options.secretsTtl : '604800000', 10) / 1e3
-    delete options.secretsTtl
-
-    const auth0Options = verifyOptions(options)
-
-    // Setup @fastify/jwt
-    instance.register(fastifyJwt, {
-      verify: auth0Options.verify,
-      cookie: options.cookie,
-      secret: getSecret,
-      jwtDecode: 'jwtDecode',
-      formatUser: options.formatUser
-    })
-
-    // Setup our decorators
-    instance.decorate('authenticate', authenticate)
-    instance.decorate('auth0Verify', auth0Options)
-    instance.decorateRequest('auth0Verify', {
-      getter: () => auth0Options
-    })
-
-    const cache =
-      ttl > 0 ? new NodeCache({ stdTTL: ttl }) : { get: () => undefined, set: () => false, close: () => undefined }
-
-    // Create a cache or a fake cache
-    instance.decorateRequest('auth0VerifySecretsCache', {
-      getter: () => cache
-    })
+    const plugin = fastifyJwtJwks(instance, options, done)
 
-    instance.addHook('onClose', () => cache.close())
+    instance.decorate('auth0Verify', instance.jwtJwks)
```

Comment:
I'm slightly unsure about this implementation, the sequence of operations doesn't look entirely right. I mean plugin initialization is asynchronous, and we're using callbacks (we should probably switch to using promises). You're initializing the other plugin but without waiting you're decorating the instance with a decorator that, as far as we know, is not yet available. 

Even if this works, I would definitely consider making it more obvious that it works.

Reply:
@simoneb I've pushed a change that adds the decorator in an onReady hook. I believe this should resolve the issue and the tests are passing.